test/k8sT/Policies.go

```diff
@@ -1122,6 +1117,183 @@ var _ = Describe("K8sPolicyTest", func() {
 			_ = kubectl.Exec(cmd)
 		})
 
+		SkipContextIf(helpers.DoesNotExistNodeWithoutCilium, "validates ingress CIDR-dependent L4 (FromCIDR + ToPorts)", func() {
+			var (
+				outsideNodeName, outsideIP string // k8s3 node (doesn't have agent running)
+
+				backendPod   v1.Pod // The pod that k8s3 node is hitting
+				backendPodIP net.IP
+
+				hostNodeName       string // Node that backendPod ends up on
+				hostIPOfBackendPod net.IP
+
+				policyVerdictAllowRegex, policyVerdictDenyRegex *regexp.Regexp
+			)
+
+			BeforeAll(func() {
+				RedeployCiliumWithMerge(kubectl, ciliumFilename, daemonCfg,
+					map[string]string{
+						"global.tunnel":               "disabled",
+						"global.autoDirectNodeRoutes": "true",
+
+						// These are required to avoid failing on net-next.
+						"global.masquerade":    "false",
+						"config.bpfMasquerade": "false",
```

Comment:
@brb For this test, I needed to do what's highlighted here. Is this a regression?

(For context, this is a new test I've added. The summary is that it's doing `(outside) host -> pod @ host A|B`, in order to validate an ingress fromCIDR + toPorts policy.)

Reply:
Hmm, the masquerading should only influence `pod @ host A|B -> outside`, but not the opposite direction flow. Did you check what happens when the masq is enabled?

Reply:
If masquerading is enabled, then what happens is the following:

```
test-output.log:{"time":"2020-07-13T20:05:10.782544267Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"SYN":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: SYN"}
test-output.log:{"time":"2020-07-13T20:05:10.782553674Z","verdict":"FORWARDED","ethernet":{"source":"3a:f7:49:1c:0d:66","destination":"16:de:a8:9e:9c:ee"},"IP":{"source":"10.0.1.109","destination":"192.168.36.13","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":80,"destination_port":46428,"flags":{"SYN":true,"ACK":true}}},"source":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"destination":{"identity":2,"labels":["reserved:world"]},"Type":"L3_L4","node_name":"k8s2","reply":true,"event_type":{"type":4,"sub_type":3},"traffic_direction":"INGRESS","Summary":"TCP Flags: SYN, ACK"}
test-output.log:{"time":"2020-07-13T20:05:10.782784247Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"RST":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: RST"}
test-output.log:{"time":"2020-07-13T20:05:11.808704554Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"RST":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: RST"}
test-output.log:{"time":"2020-07-13T20:05:13.824950432Z","verdict":"FORWARDED","ethernet":{"source":"16:de:a8:9e:9c:ee","destination":"3a:f7:49:1c:0d:66"},"IP":{"source":"192.168.36.13","destination":"10.0.1.109","ipVersion":"IPv4"},"l4":{"TCP":{"source_port":46428,"destination_port":80,"flags":{"RST":true}}},"source":{"identity":2,"labels":["reserved:world"]},"destination":{"ID":1298,"identity":43498,"namespace":"default","labels":["k8s:io.cilium.k8s.policy.cluster=default","k8s:io.cilium.k8s.policy.serviceaccount=default","k8s:io.kubernetes.pod.namespace=default","k8s:zgroup=testDS"],"pod_name":"testds-lcz5s"},"Type":"L3_L4","node_name":"k8s2","event_type":{"type":4},"traffic_direction":"INGRESS","Summary":"TCP Flags: RST"}
```

Essentially, the following happens:

1) Outside sends TCP SYN
2) Pod responds with TCP SYN ACK
3) Outside sends TCP RST (presumably because the [2] was never received)
4) The traffic is also associated with WORLD identity when it should be associated with the identity that is created from the policy.

Reply:
> 3. Outside sends TCP RST (presumably because the [2] was never received)

More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?

Reply:
> > 1. Outside sends TCP RST (presumably because the [2] was never received)
> 
> More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?

Based on the output above, I don't see how the address and port number was incorrect. Outside initiates a connection from port 46428, and all subsequent traffic uses that port. The addresses are also correct and expected.

> Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?

This sounds plausible to me. But I'm not familiar with this code, but it does sound like it aligns when masquerading is enabled.

Reply:
> > More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?
> 
> Based on the output above, I don't see how the address and port number were incorrect. Outside initiates a connection from port 46428, and all subsequent traffic uses that port. The addresses are also correct and expected.

If I understand correctly, the output above is captured at the cilium-managed node, so it may not be what the "outside" node sees. The TCP SYN is sent directly to the pod IP thanks to the route you installed on the outside node, but the reply probably goes through some SNATting in case masquerading is enabled, hence the RST. Thinking about it, it's probably expected behavior... @brb?

Reply:
> > > More likely because the TCP SYN+ACK wasn't sent to the right address or port. Otherwise, outside would just resend a TCP SYN. Maybe SYN gets SNATed, but reply SYN+ACK doesn't get reverse SNATed?
> > 
> > 
> > Based on the output above, I don't see how the address and port number were incorrect. Outside initiates a connection from port 46428, and all subsequent traffic uses that port. The addresses are also correct and expected.
> 
> If I understand correctly, the output above is captured at the cilium-managed node, so it may not be what the "outside" node sees. The TCP SYN is sent directly to the pod IP thanks to the route you installed on the outside node, but the reply probably goes through some SNATting in case masquerading is enabled, hence the RST. Thinking about it, it's probably expected behavior... @brb?

Interesting. Maybe it's worth doing a tcpdump on either end so we can get a view into what's happening. Also, fwiw, this was the same behavior I was seeing on **non**-net-next kernels with either `devices` or `host-firewall` set. I imagine if I enable those on net-next, we'll still see the same RST behavior. However, what's different here on net-next **is** the fact that I need to disable masquerading and BPF masquerading.

Reply:
Here's the tcpdump in the host namespace on the node of the backend pod, with masquerading enabled as well as BPF masquerading:

```
$     sudo tcpdump -n -i enp0s8 host 192.168.36.13
    tcpdump: verbose output suppressed, use -> or -v> for full protocol decode
    listening on enp0s8, link-type EN10MB (Ethernet), capture size 262144 bytes

             outside node +-----+                   +----------+ backend pod
                                v                   v
    22:39:46.174240 IP 192.168.36.13.36224 ^ 10.0.1.98.80: Flags [S], seq 51219120, win 64240, options [mss 1460,sackOK,TS val 1318196564 ecr 0,n                     op,wscale 7], length 0
+-->22:39:46.174598 IP 192.168.36.12.60468 > 192.168.36.13.36224: Flags [S.], seq 17444670, ack 51219121, win 64308, options [mss 1410,sackOK,TS                      val 1981994647 ecr 1318196564,nop,wscale 7], length 0
|   22:39:46.175214 IP 192.168.36.13.36224 > 192.168.36.12.60468: Flags [R], seq 51219121, win 0, length 0
|   22:39:47.181112 IP 192.168.36.13.36224 > 10.0.1.98.80: Flags [S], seq 51219120, win 64240, options [mss 1460,sackOK,TS val 1318197571 ecr 0,n                     op,wscale 7], length 0
|   22:39:47.181415 IP 192.168.36.12.60468 > 192.168.36.13.36224: Flags [S.], seq 33176448, ack 51219121, win 64308, options [mss 1410,sackOK,TS                      val 1981995654 ecr 1318197571,nop,wscale 7], length 0
|   22:39:47.182093 IP 192.168.36.13.36224 v 192.168.36.12.60468: Flags [R], seq 51219121, win 0, length 0
|
+---> This looks wrong for sure

```

Masquerading enabled, but BPF masquerading **disabled**:

```
22:53:01.770058 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [S], seq 568487496, win 64240, options [mss 1460,sackOK,TS val 1318992158 ecr 0,nop,wscale 7], length 0
22:53:01.770632 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [S.], seq 3972402295, ack 568487497, win 64308, options [mss 1410,sackOK,TS val 1982790243 ecr 1318992158,nop,wscale 7], length 0
22:53:01.771215 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [.], ack 1, win 502, options [nop,nop,TS val 1318992160 ecr 1982790243], length 0
22:53:01.771507 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [P.], seq 1:74, ack 1, win 502, options [nop,nop,TS val 1318992160 ecr 1982790243], length 73: HTTP: GET / HTTP/1.1
22:53:01.771620 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [.], ack 74, win 502, options [nop,nop,TS val 1982790244 ecr 1318992160], length 0
22:53:01.772725 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [P.], seq 1:577, ack 74, win 502, options [nop,nop,TS val 1982790245 ecr 1318992160], length 576: HTTP: HTTP/1.1 200 OK
22:53:01.773213 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [.], ack 577, win 501, options [nop,nop,TS val 1318992162 ecr 1982790245], length 0
22:53:01.778523 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [F.], seq 74, ack 577, win 501, options [nop,nop,TS val 1318992167 ecr 1982790245], length 0
22:53:01.779528 IP 10.0.1.98.80 > 192.168.36.13.36234: Flags [F.], seq 577, ack 75, win 502, options [nop,nop,TS val 1982790252 ecr 1318992167], length 0
22:53:01.779919 IP 192.168.36.13.36234 > 10.0.1.98.80: Flags [.], ack 578, win 501, options [nop,nop,TS val 1318992168 ecr 1982790252], length 0
```

So it looks like the culprit is definitely BPF masquerading on net-next. (It turns out that I can enable just masquerading, as long as BPF masquerading is disabled.)

Reply:
We can of course continue to discuss this topic wherever you folks find easiest, but it may be worth moving this discussion to a dedicated issue to make it easier to track.

Reply:
Tracked here: https://github.com/cilium/cilium/issues/12544